<a href="https://colab.research.google.com/github/ramonvluz/titanic-machine-learning-from-disaster/blob/main/Titanic_Machine_Learning_from_Disaster_Pr%C3%A1tica_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Titanic - Machine Learning from Disaster
## Prática 1 - Ramon Valgas Luz
### Considerações
-   Essa é minha primeira prática com aprendizado de maquina (Machine Learning).
-   Vou ultilizar o clássico conjunto de dados de competição do Kaggle o ["Titanic - Machine Learning from Disaster"](https://www.kaggle.com/competitions/titanic).
-   O resultado da competição é avaliado através da acuraácia.
  - "A pontuação da competição é a porcentagem de passageiros que eu consiga prevê corretamente. Essa porcentagem de acerto é conhecida como acurácia."

# Montando o Drive
- Como estou usando o [Google Colaboratory](https://colab.google/), preciso montar o drive.

  - "Montar o Google Drive no Colab permite que o código do notebook tenha acesso a todos os arquivos no Google Drive, como os conjuntos de dados que usarei nessa prática. "

In [327]:
# Montando o Drive

from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# Importando as bases
- O desafio possui dois conjuntos de dados, um para o treino do modelo e outro para realizar o teste do modelo criado para o desafio.

In [328]:
# Importando o pandas

import pandas as pd

In [329]:
# Importando e criando um objeto para a base de treino

treino = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Titanic - Machine Learning from Disaster/train.csv')

In [330]:
# Importando e criando um objeto para o conjunto de teste

teste = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Titanic - Machine Learning from Disaster/test.csv')

# Começando a entender as bases

In [331]:
# Visualisando os primeiros registros da base de treino

treino.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [332]:
# Visualisando os ultimos registros da base de treino

treino.tail(3)

PassengerId  Survived  Pclass                                      Name  \
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

In [333]:
# Agora visualisando os primeiros registros da base de teste

teste.head(3)

PassengerId  Pclass                              Name     Sex   Age  SibSp  \
0          892       3                  Kelly, Mr. James    male  34.5      0   
1          893       3  Wilkes, Mrs. James (Ellen Needs)  female  47.0      1   
2          894       2         Myles, Mr. Thomas Francis    male  62.0      0   

   Parch  Ticket    Fare Cabin Embarked  
0      0  330911  7.8292   NaN        Q  
1      0  363272  7.0000   NaN        S  
2      0  240276  9.6875   NaN        Q

In [334]:
# E também os ultimos registro da base de teste

teste.tail(3)

PassengerId  Pclass                          Name   Sex   Age  SibSp  \
415         1307       3  Saether, Mr. Simon Sivertsen  male  38.5      0   
416         1308       3           Ware, Mr. Frederick  male   NaN      0   
417         1309       3      Peter, Master. Michael J  male   NaN      1   

     Parch              Ticket     Fare Cabin Embarked  
415      0  SOTON/O.Q. 3101262   7.2500   NaN        S  
416      0              359309   8.0500   NaN        S  
417      1                2668  22.3583   NaN        C

# Analisando as bases de treino e teste

In [335]:
# Verificando as informações da base de treino

treino.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [336]:
# Verificando as informações da base de teste

teste.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


- Existem valores nulos em ambas as bases.
- A base de treino possui **12 colunas** (variáveis) e **891 entradas** (linhas).
- A base de teste possui **11 colunas** e **418 entradas**. A coluna faltante é a que preciso que o meu modelo preveja, a **Suvived**.
- Percebo também que a coluna **Survived**, na base de treino esta como **número inteiro (int64)**. Por ser um **problema de clasificação (quem morreu e quem sobreviveu / sim ou não)** acho que seria interessante tratar essa coluna e tranforma-la para o tipo **lógico/boleano (boll)**. Pesquisarei sobre e tomarei a decisão na fase de tratamento dos dados.

In [337]:
# Verificando os tipos de dados da base de treino

treino.dtypes.value_counts()

int64      5
object     5
float64    2
Name: count, dtype: int64

In [338]:
# Verificando os tipos de dados da base de teste

teste.dtypes.value_counts()

object     5
int64      4
float64    2
Name: count, dtype: int64

In [339]:
# Verificando os valores nulos na base de treino

treino.isnull().sum().sort_values(ascending=False)

Cabin          687
Age            177
Embarked         2
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
SibSp            0
Parch            0
Ticket           0
Fare             0
dtype: int64

In [340]:
# Verificando os valores nulos na base de teste

teste.isnull().sum().sort_values(ascending=False)

Cabin          327
Age             86
Fare             1
PassengerId      0
Pclass           0
Name             0
Sex              0
SibSp            0
Parch            0
Ticket           0
Embarked         0
dtype: int64

- Em ambas as bases, as variáveis (colunas) **Cabin** e **Age** possuem valores nulos.
    - Preciso tratar esses valores nas duas bases.

- A base de treino também possue 2 valores nulos na variável **Embarked**.
    - Aqui é preciso tratar esses valores, apenas na propria base de treino.

- E a base de teste possui um unico valor faltante na variável **Fare**.
    - Que também será tratado apenas na propria base teste.

# Começando os tratamentos iniciais nas bases

### Verificando a cardinalidade dos dados

In [341]:
# Verificando cardinalidade

treino.nunique().sort_values(ascending=False)

PassengerId    891
Name           891
Ticket         681
Fare           248
Cabin          147
Age             88
SibSp            7
Parch            7
Pclass           3
Embarked         3
Survived         2
Sex              2
dtype: int64

- Colunas com alta cardinalidade como **Name**, **Ticket** e **Cabin** podem atrapalhar o modelo, pois podem torná-lo menos generalizável. Nesse caso, vou eliminar essas colunas.

In [342]:
# Eliminando colunas com alta cardinalidade da base de treino

treino = treino.drop(["Name", "Ticket", "Cabin"], axis=1)

In [343]:
# Eliminando colunas com alta cardinalidade da base de teste

teste = teste.drop(["Name", "Ticket", "Cabin"], axis=1)

### Tratando os valores vazios das bases

In [344]:
# Verificando novamente os valores nulos na base de treino

treino.isnull().sum().sort_values(ascending=False)

Age            177
Embarked         2
PassengerId      0
Survived         0
Pclass           0
Sex              0
SibSp            0
Parch            0
Fare             0
dtype: int64

In [345]:
# Verificando os valores nulos na base de teste

teste.isnull().sum().sort_values(ascending=False)

Age            86
Fare            1
PassengerId     0
Pclass          0
Sex             0
SibSp           0
Parch           0
Embarked        0
dtype: int64

- **Age**
    - Por ter valores nulos em ambas as bases, irei tratar a colua **Age** primeiro.
      - E por essa coluna guardar a idade dos passageiros, decidi substituir os valores nulos pela média de todas as idades.

In [346]:
# Descobrindo a média das idades para a base de treino

treino.Age.mean()

29.69911764705882

In [347]:
# Substituindo os valores nulos pela média das idades na base de treino

treino.loc[treino.Age.isnull(), 'Age'] = treino.Age.mean()

In [348]:
# Descobrindo a média das idades para a base de teste

teste.Age.mean()

30.272590361445783

In [349]:
# Substituindo os valores nulos pela média das idades na base de teste

teste.loc[teste.Age.isnull(), 'Age'] = teste.Age.mean()

- **Embarked**
  - Na base de **treino** ainda tenho 2 valores nulos na coluna **Embarked**.
    - Vou verificar todos os valores nessa coluna e substituir os valores nulos pela **moda** da coluna.

In [350]:
# Verificando os valores na coluna Embarked na base de treino

treino.Embarked.value_counts()

Embarked
S    644
C    168
Q     77
Name: count, dtype: int64

In [351]:
# Selecionando a moda dessa coluna

treino.Embarked.mode()[0]

'S'

In [352]:
# Substituindo os valores vazios pela moda na coluna Embarked na base de treino

treino.loc[treino.Embarked.isnull(), 'Embarked'] = treino.Embarked.mode()

- **Fare**
  - Na base de teste ainda possuo um unico valor nulo na coluna **Fare**.
  - Como essa coluna se trata do valor da tarifa paga pelo passageiro, vou ultilizar a mesma lógica usada na coluna **Age** e vou substituir esse valor pela média das tarifas.

In [353]:
# Descobrindo o valor médio das tarifas na base de teste

teste.Fare.mean()

35.627188489208635

In [354]:
# Substituindo os valores vazios pela moda na coluna Fare na base de teste

teste.loc[teste.Fare.isnull(), 'Fare'] = teste.Fare.mean()

- **Verificando os valores nulos novamente para confirmar que foram devidamente tratados**.

In [355]:
# verificando os valores nulos da base de treino

treino.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Sex            0
Age            0
SibSp          0
Parch          0
Fare           0
Embarked       2
dtype: int64

In [356]:
# verificando os valores nulos da base de teste

teste.isnull().sum()

PassengerId    0
Pclass         0
Sex            0
Age            0
SibSp          0
Parch          0
Fare           0
Embarked       0
dtype: int64

## Considerando apenas as colunas que não são de texto

- Os algoritimos que irei usar não aceitam variáveis de texto, poir isso vamos tratar a base para usarmos apenas as colunas que não são de texto.

In [357]:
# Verificando as colunas de não são de texto na base de treino

col_treino_nr = treino.columns[treino.dtypes != 'object']
col_treino_nr

Index(['PassengerId', 'Survived', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare'], dtype='object')

In [358]:
# Selecionando apenas os valores numéricos da base de treino

treino_nr = treino.loc[:, col_treino_nr]

In [359]:
# Verificando as colunas de não são de texto na base de teste

col_teste_nr = teste.columns[teste.dtypes != 'object']
col_teste_nr

Index(['PassengerId', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare'], dtype='object')

In [360]:
# Selecionando apenas os valores numéricos da base de teste

teste_nr = teste.loc[:, col_teste_nr]

# Selecionando um modelo para classificar os dados
- Para começar, vou testar os seguintes modelos:
    - **Árvore de classificação**
        - https://scikit-learn.org/stable/modules/tree.html#classification
    - **Classificação dos vizinhos mais próximos**
        - https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html#sklearn.neighbors.KNeighborsClassifier
    - **Regressão Logística**
        - https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html#sklearn.linear_model.LogisticRegression

- Antes de usar os algoritmos, preciso dividir a base de treino em **treino e validação**.
    - Para isso vou utilizando o **train_test_split**.
        - https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html

In [361]:
# Importando o train_test_split

from sklearn.model_selection import train_test_split

In [362]:
# Separando a base de treino em X e y

X = treino_nr.drop(['PassengerId','Survived'], axis=1)
y = treino.Survived

- Retirei a coluna **PassengerID** pois ela não é relevante para o aprendizado do modelo.
- E tambeém retirei a coluna **Suvived**, pois é a coluna quê eu quero prever.

Observação: Não removi a coluna **Survived** durante as etapas iniciais de tratamento de dados. Apenas a elimino nesta fase, pois, ao enviar o arquivo para o Kaggle a fim de avaliar meu modelo e obter minha pontuação, é necessário incluir esta coluna, juntamente com a coluna **PassengerID**.

In [363]:
# Separando em treino e validação

X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.33, random_state=13)

- **Árvore de classificação**

In [364]:
# Importando o modelo de árvore de classifucação

from sklearn import tree

In [365]:
# Criando o classificador

clf_ac = tree.DecisionTreeClassifier()

In [366]:
# Fazendo o fit dos dados

clf_ac = clf_ac.fit(X_train, y_train)

In [367]:
# Fazendo a previsão

y_pred_ac = clf_ac.predict(X_val)

- **KNeighborsClassifier**

In [368]:
# Importando o modelo

from sklearn.neighbors import KNeighborsClassifier

In [369]:
# Criando o classificador

clf_knn = KNeighborsClassifier(n_neighbors=3)

In [370]:
# Fazendo o fit dos dados

clf_knn = clf_knn.fit(X_train,y_train)

In [371]:
# Fazendo a previsão

y_pred_knn = clf_knn.predict(X_val)

- **Regressão Logística**

In [372]:
# Importando o modelo

from sklearn.linear_model import LogisticRegression

In [373]:
# Criando o classificador

clf_rl = LogisticRegression(random_state=42)

In [374]:
# Fazendo o fit dos dados

clf_rl = clf_rl.fit(X_train,y_train)

In [375]:
# Fazendo a previsão

y_pred_rl = clf_rl.predict(X_val)

## Avaliando os modelos
- Para fazer essa análise, irei usar:
    - Acurácia (método de avaliação usado na competição):
        - https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html
    - Matriz de confusão (ajuda a visualizar a distribuição dos erros):
        - https://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html

-  Avaliando a **acurácia**

In [376]:
# Importando

from sklearn.metrics import accuracy_score

In [377]:
# Para a árvore de clasificação

accuracy_score(y_val, y_pred_ac)

0.688135593220339

In [378]:
# Para o knn - KNeighborsClassifier

accuracy_score(y_val, y_pred_knn)

0.6542372881355932

In [379]:
# Para a regressão logística

accuracy_score(y_val, y_pred_rl)

0.7288135593220338

- Avaliando a **matriz de confusão**

In [380]:
# Importando

from sklearn.metrics import confusion_matrix

In [381]:
# Para a árvore de clasificação

confusion_matrix(y_val, y_pred_ac)

array([[139,  48],
       [ 44,  64]])

In [382]:
# Para o knn - KNeighborsClassifier

confusion_matrix(y_val, y_pred_knn)

array([[134,  53],
       [ 49,  59]])

In [383]:
# Para a regressão logística

confusion_matrix(y_val, y_pred_rl)

array([[172,  15],
       [ 65,  43]])

- O modelo de Regressão Logística (clf_rl) destacou-se com a maior acurácia de **0.72** entre os três modelos avaliados.
    - Sua superioridade deve-se à capacidade de generalização nos dados de validação, refletindo em uma acurácia mais alta. Reconhecida pela simplicidade e eficiência em problemas de classificação binária, como a previsão de sobrevivência no Titanic, a Regressão Logística também demonstrou ser menos sensível a outliers e overfitting em comparação aos outros modelos testados, especialmente quando os dados são tratados e balanceados adequadamente.

# Fazendo a previsão para os dados de teste
- Vou ultilizar o modelo com melhor precisão para fazer o predição na base de teste. No caso a regressão logistica.

In [384]:
# Visualizando o X_train

X_train.head(3)

Pclass   Age  SibSp  Parch      Fare
791       2  16.0      0      0   26.0000
318       1  31.0      0      2  164.8667
781       1  17.0      1      0   57.0000

In [385]:
# Visualizando a base de teste (apenas com valores numéricos)

teste_nr.head(3)

PassengerId  Pclass   Age  SibSp  Parch    Fare
0          892       3  34.5      0      0  7.8292
1          893       3  47.0      1      0  7.0000
2          894       2  62.0      0      0  9.6875

In [386]:
# Para a base de teste ser igual a base de treino, preciso eliminar a coluna de id

X_teste = teste_nr.drop("PassengerId",axis=1)

In [387]:
# Utilizando a regressão logística na base de teste

y_pred = clf_rl.predict(X_teste)

# Preparando a base para o envio no Kaggle.

In [388]:
# Criando uma nova coluna com a previsão na base de teste

#teste['Survived'] = y_pred

In [389]:
# Selecionando apenas a coluna de Id e Survived para fazer o envio

#base_envio = teste[['PassengerId','Survived']]

In [390]:
# Exportando para um csv

# base_envio.to_csv('resultados2.csv',index=False)
